framework/sysdeps/windows/tmpfile.c

```diff
@@ -40,9 +41,68 @@ static unsigned next_random()
     return result;
 }
 
+static bool is_dir_writable(const wchar_t *dir)
+{
+    if (!dir)
+        return false;
+
+    size_t len = wcslen(dir);
+    if (!len)
+        return false;
+    if (dir[len - 1] == L'\\')
+        return _waccess(dir, W_OK) == 0;
+
+    // append a backslash so we confirm it's a directory
+    wchar_t path[MAX_PATH];
+    wcscpy(path, dir);
+    wcscat(path, L"\\");
+    return _waccess(path, W_OK) == 0;
+}
+
+static const wchar_t *find_alternate_temp_dir()
+{
+    static const wchar_t *dirname = NULL;
+    if (dirname)
+        return dirname;
+
+    dirname = _wgetenv(L"TEMP");
+    if (is_dir_writable(dirname))
+        return dirname;
+
+    dirname = _wgetenv(L"TMP");
+    if (is_dir_writable(dirname))
+        return dirname;
+
+    return NULL;
+}
+
+// fills in the temporary directory in tmpname and returns the length of the path
+// (including the terminating backslash)
+static size_t fill_tmpdir(wchar_t tmpname[MAX_PATH])
+{
+    size_t len = GetTempPathW(MAX_PATH, tmpname);
+    if (len) {
+        wcscat(tmpname, L"\\");
+        if (is_dir_writable(tmpname))
+            return len;
+    }
+
+    const wchar_t *othertmp = find_alternate_temp_dir();
+    if (!othertmp) {
+        // use the CWD as a last resort
+        wcscpy(tmpname, L".\\");
+        return 2;
+    }
+
+    wcscpy(tmpname, othertmp);
+    wcscat(tmpname, L"\\");
+    return wcslen(tmpname);
+}
+
 int open_memfd(enum MemfdCloexecFlag flag)
 {
-    wchar_t tmpname[sizeof SANDSTONE_STRINGIFY(UINT_MAX) ".tmp"];
+    wchar_t tmppath[MAX_PATH];
+    wchar_t *tmpname = tmppath + fill_tmpdir(tmppath);
```

Comment:
If `fill_tmpdir`, and its underlying API calls, return string with length longer than `MAX_PATH`, then `tmpname` pointer would exceed boundaries of `tmppath` array. 
We're trying to append name of a temp file few lines below without checking if the name will fit into the buffer.
Would be best to reserve at least 12 characters in the `tmppath` and - if total length of temp directory path + temp file name - exceeded `MAX_PATH - 12` - convert name of the file to 8.3 format.

Reply:
Fixing.